Order manager

DmitryIldyukov opened the pull request.

DmitryIldyukov committed:
1 lab

DmitryIldyukov committed:
Теперь показываться только дата доставки, без времени

DmitryIldyukov committed:
Добавил проверку на пустой ввод

DmitryIldyukov committed:
Удалил лишний using

DmitryIldyukov committed:
Добавил editorconfig

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,88 @@
+    }
+}
+
+public class Order
```
давай класс вытащим в отдельный файл)

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,88 @@
+
+public class Order
+{
+    public Order( string productName,
```
давай в одну строчку, вроде как нормально умещается:
```c#
    public Order( string productName, int quantity, string customerName, string address )
```

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,88 @@
+        Address = address;
+    }
+
+    public string ProductName { get; private set; }
```
Обычно свойства пишутся выше конструктора)

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,88 @@
+            Console.Write( message );
+            input = Console.ReadLine();
+
+            if ( string.IsNullOrWhiteSpace( input ) )
```
у if лучше указывать фигурные скобочки, даже если логика под условием небольшая)

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,88 @@
+    static void OrderDelivery( Order order )
+    {
+        Console.WriteLine( $"{order.CustomerName}! Ваш заказ {order.ProductName} в количестве {order.Quantity} оформлен! " +
+                $"Ожидайте доставку по адресу {order.Address} к {DateTime.UtcNow.AddDays( 3 ).ToLocalTime().ToShortDateString()}" );
```
``` c#
{DateTime.UtcNow.AddDays( 3 ).ToLocalTime().ToShortDateString()}
```

не очень хорошо выполнять подобные преобразования, потому как передавая объект, которые содержит информацию, лучше чтобы он уже имел достоверные данные, которые  необходимо только вывести

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,88 @@
+
+        Console.Write( "Введите количество товара: " );
+        int quantity;
+        while ( !int.TryParse( Console.ReadLine(), out quantity ) || quantity <= 0 )
```
раз у тебя есть отдельная функция для валидации ввода строкового типа, то можно и написать функцию для валидации входных числовых данных, так будет смотреться лаконичнее)

Nafanyan replied:
ещё лучше ставить фигурные скобочки)

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,88 @@
+        OrderCancelation( order );
+    }
+
+    // Вынес в отдельный метод, т.к. в теории тут будет выполняться множество других методов
```
Комментарии, за исключением особенностей бизнес логики, избыточны, но спасибо я ценю твоё желание облегчить мне понимание твоего кода)

Только немного не понял "в теории тут будет выполняться множество других методов", где тут? каких методов?

DmitryIldyukov replied:
Решил добавить комментарий, т.к. подумал, что выносить в отдельный метод вывод в консоль одной строки не совсем правильное решение (т.к. этот метод используется лишь в одном месте) Тем не менее все же вынес, потому что, если продумывать логику доставки далее, то в этих методах будут еще другие методы, например, сборка заказа)

Nafanyan reviewed (changes requested):
Круто!

Хорош, попили всё на методы, выделил отдельный класс для хранения информации о заказе, но есть некоторые замечания)

DmitryIldyukov committed:
Исправление замечаний

1) Поправил форматирование
2) Убрал лишние комментарии
3) Вынес класс Order в отдельный файл
4) Переписал цикл do-while на обычный while
5) Добавил метод для получения кол-ва продукта
6) Дата доставки хранится в Order

DmitryIldyukov committed:
Решил хранить дату доставки в UTC формате

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,85 @@
+        return Console.ReadLine().ToLower() == "y";
+    }
+
+    static Order CreateOrder()
```
Скорее придирка, давай размещать методы в порядке их вызова, чем раньше вызывается метод, тем  выше он будет стоять, так иерархия будет следующая:
CreateOrder()
GetNonEmptyStringFromConsole
GetValidProductQuantity
....
OrderCancelation

Nafanyan commented on OrderManager/Program.cs:
```diff
@@ -0,0 +1,85 @@
+
+        string address = GetNonEmptyStringFromConsole( "Введите адрес доставки: ", "Неверный ввод. Адрес доставки не может быть пустым. Попробуйте снова." );
+
+        return new Order( productName: productName, quantity: quantity, customerName: customerName, address: address );
```
именованные аргументы это иногда необходимо, но здесь как будто идет дублирование, давай уберем)

Имена переменных итак описывают входные данные, необходимые для создания объекта Order

Nafanyan reviewed (approved):
Всё супер, только мелкие недочеты)

Прежде чем мержить сделай squash, это специальный механизм для объединения нескольких коммитов в один. Будет неприятно если у тебя в main попадет 6 коммитов исправлений по одной задаче, лучше пусть будет один)

Насколько я знаю в гитхабе это можно сделать выбрав
![image](https://github.com/DmitryIldyukov/TLPractice/assets/90004716/97b9cc29-6853-4e64-acca-7cd8af7d5608)

DmitryIldyukov committed:
Расположил методы в порядке выполнения
